### ISSUE TITLE
SIDE_MODULE=2 deletes functions of exported class that are used and called

### ISSUE THREAD

**Issue by Maksim-Bozhko** [emscripten-core/emscripten#16762]
Steps to reproduce: 
// 1.h
```
class __attribute__((visibility("default"))) A
{
public:
	int foo();
};
```
// 1.cpp
```
#include "1.h"
int A::foo() {
	return 123;
}
```
emcc 1.cpp -fvisibility=hidden -sSIDE_MODULE=2 -o out/1.wasm
// main.cpp
```
#include "1.h"
#include <stdio.h>
int main() {
	A a;
	printf("%d\n", a.foo());
	return 0;
}
```
emcc main.cpp out/1.wasm -sMAIN_MODULE=2 -o out/main.html

This results in linking error complaining about A::Foo() not being defined. Yes I know that EMSCRIPTEN_KEEPALIVE on function declaration will fix this, but we are currently porting very large(8.5 million LOC) project that has lots of exported classes and marking every function declaration with EMSCRIPTEN_KEEPALIVE seems insane. Why does dead code elimination kill member functions of an exported class to begin with? The usual C++ practice is to mark all symbols of dynamic library hidden by default, then explicitly mark classes we want to export with special compiler attributes. We need dead code elimination to reduce size of dynamic libraries, but deleting exports is very unexpected. Is there a way to stop dead code elimination from removing exports?

**Comment by sbc100** [comment]
I believe that if you want all the default symbols to be exported you would need to use `-sSIDE_MODULE=1`. 

`-sSIDE_MODULE=2` limits exports to those explicitly specified via the command line or via `EMSCRIPTEN_KEEPALIVE`.

**Comment by Maksim-Bozhko** [comment]
@sbc100 thank you, but is there a way to have dead code elimination that doesn't remove exports? -O3 I guess? -Wl,-dead_strip maybe?

**Comment by sbc100** [comment]
Yes, DCE will still happen in the wasm-ld linker.   Code and data will not be kept that is not reachable by any of your exports.

**Comment by Maksim-Bozhko** [comment]
Thank you. Related question: it seems that .wasm dynamic library is not as self-sufficent as windows dll for example. In a sense that it doesn't understand transitive dependencies and final linking of executale needs collecting all deps of deps. Is there any version of dynamic library that is more complete, or my understanding is wrong?

**Comment by sbc100** [comment]
I'm not quite sure I understand your question but I'll take a stab at answering.

Yes, the SIDE_MODULE/MAIN_MODULE system differs in some ways from the way in which shared libraries work on windows and linux.   Side modules will always succeed in linking even if they contain undefined symbols, which means you don't need to specify their dependencies on the command line.. however if you do specify those dependencies they should be included in the binary.   You can use wasm-objdump to see them listed:

```
$ ./emcc tests/hello_world.c -o other_side.wasm -sSIDE_MODULE=2 side.wasm
$ wasm-objdump -x other_side.wasm
...
Custom:
 - name: "dylink.0"
 - mem_size     : 15
 - mem_p2align  : 0
 - table_size   : 0
 - table_p2align: 0
 - needed_dynlibs[1]:
  - side.wasm
...
```

Here you can see side.wasm listed as a dependency of other_side.wasm..

When performing the final link with `MAIN_MODULE=2`, emcc will traverse the transitive dependency graph of all side modules so that any symbols from the main libraries that are needed will get included.  See: https://github.com/emscripten-core/emscripten/blob/54675bac246e84ca024c182e477665d21fe2e2f7/emcc.py#L671-L687

Having said that the system is different to linux and windows dynamic linking and its not unlikely you will run into other behavioral differences.
